p2p/discover/v4_udp.go

```diff
@@ -750,7 +831,41 @@ func decodeV4(buf []byte) (packetV4, encPubkey, []byte, error) {
 	return req, fromKey, hash, err
 }
 
-// Packet Handlers
+// checkBond checks if the given node has a recent enough endpoint proof.
+func (t *UDPv4) checkBond(id enode.ID, ip net.IP) bool {
+	return time.Since(t.db.LastPongReceived(id, ip)) < bondExpiration
+}
+
+// ensureBond solicits a ping from a node if we haven't seen a ping from it for a while.
+// This ensures there is a valid endpoint proof on the remote end.
+func (t *UDPv4) ensureBond(toid enode.ID, toaddr *net.UDPAddr) {
+	tooOld := time.Since(t.db.LastPingReceived(toid, toaddr.IP)) > bondExpiration
+	if tooOld || t.db.FindFails(toid, toaddr.IP) > 5 {
+		rm := t.sendPing(toid, toaddr, nil)
+		<-rm.errc
+		// Wait for them to ping back and process our pong.
+		time.Sleep(respTimeout)
```

Comment:
Blind sleep? Seems a bit odd? Shouldn't we perhaps wake up as soon as the rely arrives and not wait until we reach the timeout?

Reply:
We had what you describe before https://github.com/ethereum/go-ethereum/commit/f0c6f921408d97251b1b554a7ea83d020c70a2da#diff-67724ae2da807c2b2c15ec4cd9cd4c6eR341, but it was unsafe in case of packet reordering. The expectation here is that we send a ping to the remote node to make it ping back. When it does that, we need to reply to the ping first, then wait a bit, then send findnode.